### PR TITLE
IGNITE-22278 Sql. Refactor base and planning contexts

### PR DESCRIPTION
**Pull request by korlov42**
https://issues.apache.org/jira/browse/IGNITE-22278

This patch refactors contexts used within sql engine:
1) BaseQueryContext is removed. Instead lightweight SqlOperationContext is introduced. All planning-related metadata moved to PlanningContext
2) PrepareService is not aware of SchemaManager. It accepts operation time as input and resolve active schema itself
3) `schemaVersion` renamed to `catalogVersion` in numerous places to better reflect semantic

-------------------

Thank you for submitting the pull request.

To streamline the review process of the patch and ensure better code quality
we ask both an author and a reviewer to verify the following:

### The Review Checklist
- [ ] **Formal criteria:** TC status, codestyle, mandatory documentation. Also make sure to complete the following:  
\- There is a single JIRA ticket related to the pull request.  
\- The web-link to the pull request is attached to the JIRA ticket.  
\- The JIRA ticket has the Patch Available state.  
\- The description of the JIRA ticket explains WHAT was made, WHY and HOW.  
\- The pull request title is treated as the final commit message. The following pattern must be used: IGNITE-XXXX Change summary where XXXX - number of JIRA issue.
- [ ] **Design:** new code conforms with the design principles of the components it is added to.
- [ ] **Patch quality:** patch cannot be split into smaller pieces, its size must be reasonable.
- [ ] **Code quality:** code is clean and readable, necessary developer documentation is added if needed.
- [ ] **Tests code quality:** test set covers positive/negative scenarios, happy/edge cases. Tests are effective in terms of execution time and resources.

### Notes
- [Apache Ignite Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Java+Code+Style+Guide)

### CODE REVIEW

**Comment by lowka** [comment]
There are few places in sql-engine module where `schemaVersion` is still being used:

- `KeyValueModifyPlan` has a field `schemaVersion`
- `KeyValueGetPlan` has a field `schemaVersion`
- `SqlSchemaManager` defines a method `table(schemaVersion, tableId)`
- `ExecutionServiceImpl FragmentCacheKey` has a field `schemaVersion`

Are the names of these fields/variables kept unchanged on purpose?

**Comment by korlov42** [comment]
@lowka my bad, thanks for noticing. I didn't use full text search, just changed names in those places I stumbled upon.

Renamed also mentioned places.

**Review by lowka — APPROVED** [review]

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/schema/IgniteSchema.java [hunk 1]
```diff
@@ -44,7 +44,7 @@ public class IgniteSchema extends AbstractSchema {
     /** Constructor. */
     public IgniteSchema(String name, int version, Collection<? extends IgniteDataSource> tables) {
         this.name = name;
-        this.version = version;
+        this.catalogVersion = version;
```
Comment (xtern) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/schema/IgniteSchema.java:47]: Perhaps it makes sense to rename the parameter too?
(and the SCHEMA_VERSION constant in the ExecutionServiceImplTest test)

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/schema/SqlSchemaManager.java [hunk 2]
```diff
@@ -25,23 +25,23 @@
  */
 public interface SqlSchemaManager {
     /**
-     * Returns root schema of the given version.
+     * Returns root schema derived from catalog of the given version.
      */
     SchemaPlus schema(int version);
```
Comment (xtern) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/schema/SqlSchemaManager.java:30]:
```suggestion
    SchemaPlus schema(int catalogVersion);
```

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/PrepareServiceImpl.java [hunk 3]
```diff
@@ -126,7 +133,8 @@ public static PrepareServiceImpl create(
             DataStorageManager dataStorageManager,
             MetricManager metricManager,
             SqlDistributedConfiguration clusterCfg,
-            SqlLocalConfiguration nodeCfg
+            SqlLocalConfiguration nodeCfg,
+            SqlSchemaManager schemaManager
```
Comment (xtern) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/PrepareServiceImpl.java:137]: javadoc is missing :sunglasses:

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/PrepareServiceImpl.java [hunk 4]
```diff
@@ -156,17 +165,18 @@ public PrepareServiceImpl(
             DdlSqlToCommandConverter ddlConverter,
             long plannerTimeout,
             int plannerThreadCount,
-            MetricManager metricManager
+            MetricManager metricManager,
+            SqlSchemaManager schemaManager
```
Comment (xtern) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/PrepareServiceImpl.java:169]: javadoc is missing

**Review by xtern — APPROVED** [review]

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/schema/IgniteSchema.java (hunk 1 above)
Comment (korlov42) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/schema/IgniteSchema.java]: done

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/schema/SqlSchemaManager.java (hunk 2 above)
Comment (korlov42) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/schema/SqlSchemaManager.java]: renamed

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/PrepareServiceImpl.java (hunk 3 above)
Comment (korlov42) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/PrepareServiceImpl.java:138]: fixed, thanks!

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/PrepareServiceImpl.java (hunk 4 above)
Comment (korlov42) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/PrepareServiceImpl.java:171]: done

**Review by xtern — APPROVED** [review]

**Pull request merged**